src/main/java/org/mastodon/mamut/fitting/edgel/SampleEllipsoidEdgel.java

```diff
@@ -60,91 +64,119 @@ public static Map< Edgel, Double > getCosts(
-				final double[] radii = ellipsoid.getRadii();
-				if ( Double.isNaN( radii[ 0 ] ) || Double.isNaN( radii[ 1 ] ) || Double.isNaN( radii[ 2 ] ) )
-				{
-					--sample;
-					continue;
-				}
-
-				ellipsoid.localize( center );
-				LinAlgHelpers.subtract( center, expectedCenter, center );
-				if ( LinAlgHelpers.length( center ) > maxCenterDistance )
-					continue;
-
-				final double cost = costFunction.compute( ellipsoid, edgels );
-				if ( cost < bestCost )
-				{
-					bestCost = cost;
-					bestEllipsoid = ellipsoid;
-				}
-			}
-			catch ( final IllegalArgumentException e )
-			{
-//				e.printStackTrace();
-//				System.out.println( "oops" );
-			}
-			catch ( final RuntimeException e )
+			sampleCoordinatesFromEdgels( edgels, coordinates, rand, indices );
+
+			final Ellipsoid ellipsoid = tryFitEllipsoidYuryPetrov( coordinates );
+
+			if ( ! isEllipsoidValid( ellipsoid, expectedCenter, maxCenterDistance, center ) )
+				continue;
+
+			final double cost = costFunction.compute( ellipsoid, edgels );
+			if ( cost < bestCost )
 			{
-//				System.out.println( "psd" );
+				bestCost = cost;
+				bestEllipsoid = ellipsoid;
 			}
+
+			if( ++candidates >= numCandidates )
+				break;
 		}
 
-		if ( bestEllipsoid == null )
-			return null;
+		if ( bestEllipsoid == null ) // no ellipsoid found
+			throw new NoEllipsoidFoundException( "No ellipsoid found, that is near to the expected center." );
 
-		final Ellipsoid refined = fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance, angleCutoffDistance );
-		if ( refined == null )
+		try
+		{
+			// refined ellipsoid
+			return fitToInliers( bestEllipsoid, edgels, outsideCutoffDistance, insideCutoffDistance,
+					angleCutoffDistance );
+		}
+		catch ( final RuntimeException e )
 		{
-//			System.err.println( "refined ellipsoid == null! This shouldn't happen!");
+			// refinement failed
 			return bestEllipsoid;
 		}
-		return refined;
+	}
+
+	private static void sampleCoordinatesFromEdgels( List< Edgel > edgels, double[][] coordinates, Random rand, TIntArrayList indices )
+	{
+		indices.clear();
+		for ( int s = 0; s < coordinates.length; ++s )
+		{
+			int i = rand.nextInt( edgels.size() );
+			while ( indices.contains( i ) )
+				i = rand.nextInt( edgels.size() );
+			indices.add( i );
+			edgels.get( i ).localize( coordinates[ s ] );
+		}
+	}
+
+	private static Ellipsoid tryFitEllipsoidYuryPetrov( double[][] coordinates )
+	{
+		try
+		{
+			return FitEllipsoid.yuryPetrov( coordinates );
+		}
+		catch ( final RuntimeException e )
+		{
+			return null;
+		}
+	}
+
+	private static boolean isEllipsoidValid( Ellipsoid ellipsoid, double[] expectedCenter, double maxCenterDistance, double[] center )
+	{
+		if ( ellipsoid == null )
+			return false;
+
+		// skip degenerate samples
+		final double[] radii = ellipsoid.getRadii();
+		if ( Double.isNaN( radii[ 0 ] ) || Double.isNaN( radii[ 1 ] ) || Double.isNaN( radii[ 2 ] ) )
+			return false;
+
```

Comment:
```suggestion

		// ellipsoid with negative radii
		if ( radii[ 0 ] <= 0 || radii[ 1 ] <= 0 || radii[ 2 ] <= 0 )
			return false;
```
One could add this validation check so that only ellipsoid with positive-only radii are considered valid.

Reply:
I went for a slightly different solution. I added a method `ellipsoid.isLegitimate()` that tests if all the radii of  the ellipsoid are strictly positive. I used this method in the if condition.

Reply:
Okay. Please resolve.